readme.txt
=== Event Tickets with Ticket Scanner ===
Contributors: Saso Nikolov
Tags: events, party tickets, ticket scanner, redeem tickets, woocommerce, venue tickets
Requires PHP: 7.0
Stable tag: 1.4.7
Tested up to: 6.2
License: GPLv3
License URI: https://www.gnu.org/licenses/gpl-3.0.html

Sell tickets for your event or party with WooCommerce. Scan the tickets at the entrance.
You can scan the tickets of your customer from the QR code on their mobile devices or the printed PDF ticket.

== Description ==
Selling event or party tickets and redeeming the tickets was never so easy.
https://youtu.be/lqY6Yh6CHJ4
[Quick start video](https://youtu.be/lqY6Yh6CHJ4)

Sell tickets for your event or party and redeem the ticket at the entrance with the free Tickets with WooCommerce plugin, from the team behind the Serial Code Validator in WordPress.

You can make your tickets look unique by adding a banner, logo and background image. The same applies also for the event or party flyer.

This plugin makes it easy to sell tickets and redeem them at the event location entrance for you in-person or virtual events. Easily integrate event tickets with your WooCommerce products.

Install the plugin and set your products to be a event ticket product. That's it.

The plugin will generate a secure ticket number and add them to the mail to your clients. It will also generate a PDF with the ticket information and QR code. The built-in ticket scanner allows you and your team to redeem the tickets by scanning the QR-code. You can also use an external keyboard scanner by using the text input field on the ticket scanner page.

You can even import your own ticket numbers, but it will generate them also on the fly. The freedom is in your hands.

**Shop:** Sell event and party tickets with your normal products. The plugin will create the ticket number on the fly or use unused ticket numbers from your ticket number list for each sold item.

**Validation of ticket numbers:** You customer can validate a ticket number on your page. The ticket number generator gives you a flexible pattern to generate ticket numbers upfront. You can also **import your existing codes**.

**Validation of tickets:** You customer will get a link to a detail page and a PDF for each sold ticket.

**Build your own stolen ticket number database:** Support your customer and yourself to protect stolen ticket numbers to be sold or used to others. Mark reported ticket numbers as stolen and let others check the ticket numbers with the ticket number validator.

**How can you setup the ticket numbers?**
The ticket numbers will be generated by the plugin automatically, if you do not set up ticket numbers upfront.
You can enter your ticket numbers within the WP admin area in the ticket number section (a new menu point in the wordpress amdin area).
You can generate your own ticket numbers manually using the ticket number generator or import existing ticket numbers.
And you are also able to add your own ticket numbers by just entering them to the generated ticket numbers below the ticket number generator.
If needed you can assign the ticket (numbers) to a list which helps you to organize the tickets. The ticket list need to be assigned to a WooCommerce product.
Create one or more ticket lists to use the WooCommerce functionality.

**How the ticket number works**
You define your look and feel for your ticket numbers. The generated ticket numbers will be unique. After you are satisfied with the generated ticket numbers, you can start the adding step.
The ticket adding step will check if the ticket number is unique on the server before adding it.
If the ticket number exists already on the server, it will not be added.

= The ticket number generator provides a lot of features: =
- Ticket number prefix - you can enter a fixed prefix to each ticket number
- Length of ticket number
- Only letters
- Uppercase, Lowercase or both
- Exclusion of letters that are easy to missread (i,l,o,q,p)
- Delimiter possible: [None, -, :, SPACE] - you decide after how many letters within the ticket number
- CVV code generation (only numbers)

Yes, this is a fork of the serial code validator plugin. We decided to make it an own plugin out of it. So to add more and more features that are not related to serial codes but to event tickets.

== Features ==
At the bottom of this page you will find a complete list of all features.

https://youtu.be/ls_Lkf08n9I

* Sell event tickets as PDF with WooCommerce
* Download PDF of ticket to print it as a badge for your customers
* Attach your own PDF to the ticket PDF (will be added as additional pages)
* Also for professional usage - use your QR and barcode scanner to verify the tickets
* Product variants supported
* Generate flyer for your party or event
* Redeem event tickets at the entrance using the included ticket scanner page (mobile ready)
* Add ICS calendar file or a ticket to the purhcase email and ticket detail page
* Store WooCommerce orderid, itemid and productid to a ticket that was generated or used for a product sale
* You can now set a unique ticket number format for all WooCommerce product that are using a ticket number
* You can now set the ticket number format directly also on the WooCommerce product detail page if needed
* Use your codes to restrict purchases that allow a purchase of this product only if the buyer has a code for it (purchase allowance code)
* Add your own messages for the ticket number validation form for your customers
* Add your own message for the "product stolen" validation message
* Disable the validation form for not logged in wordpress user
* User can register to a ticket (with the wordpress user id if needed) after the ticket number is checked - this makes your code one-time usable
* Display registered user information of a ticket number during the validation if you need this
* One time check can have a maximum check amount based on ticket list or based on the global settings
* The user can be forwared (redirected) to an URL after the ticket number was checked - to show more details
* Webhooks - you can inform other systems about ticket redeemed status and validation steps
* Display assigned tickets to your user with a shortcode [sasoEventTickets_code]
* Add images to the ticket (header, background and Footer)
* Adopt font size for the PDF ticket
* Forcing responsive design for the ticket scanner for better experience
* Allow multiple redeem times for multi usage tickets


== GETTING STARTED ==

A Quick start is shown in this [video](https://vollstart.de/event-tickets-with-ticket-scanner/docs/#quickstart).
A good first start is to open the event ticket admin area and create a list first, if not already done or if you do not want to use the default ticket list.
Go to your WooCommerce product and activate the ticket sale option and set the list.
Check out all the possible options in the event ticket admin area to understand, how to tweak your usages of plugin.
Optional: Then add your ticket number by importing (add button at the ticket table) or assign the ticket list to your products.
Optional: If you need a validation form for your users, to check the ticket number, then please add the shortcode **[sasoEventTicketsValidator]** to a page.

= Steps to start =
- Go to the admin area and click on menu "Event Tickets".
- Click on button "Add" next to the heading "List".
- Go to your "ticket" product and set the option with in the product settings "Event Tickets"
*To scan the QR code of the sold tickets at the entrance:*
- Go to the admin area and click on menu "Event Tickets".
- Click on the button "Ticket Scanner" at the top area
- Scan tickets and redeem them

== Premium Features ==
* Unlimited ticket numbers (the basic allows you 50 tickets)
* Unlimited Lists
* Save IP address of user using the ticket number validation check on your page
* De- and re-activate tickets
* Block IPs after x retries within 60 minutes - prevent brute force
* One time usage overuled by list settings
* Enter expiration date for tickets
* Enter expiration date for ticket number lists (ticket number overrule list expiration date)

Try it out first, before you go Premium. ["Here you can find the premium plugin"](https://vollstart.de/event-tickets-with-ticket-scanner).

== Build your own Stolen products database ==
You can set the status of a ticket number to be marked as stolen. This way you can offer your customers a way to report a stolen product. Other buyers of used products can validate the ticket number on your website and will be notified with a "Product is stolen" warning.
This could increase sales and help reducing theft of your products.
The check behaviour will be like "inactive codes" but with the stolen message. You can change the stolen message within the options area of the serial code admin.

== Protection with CVV ==
You can set a code verification value (CVV) to protect your tickets. The validation check will ask the user to enter both for a valid check.
That way you increase the security of your ticket numbers and prevent stealing the numbers (If you are able hide the CVV on your packaging).

changelog.txt
= 1.4.0 - 2023-05-01 =
* Ticket badge PDF Editor for your customer. Use it as a name sign or venue badge
* Ticket badge PDF download for the ticket
* Each option display now the option key, that can be used as a placeholder on the ticket badge designer
* Clean up additional output before the ticket PDF is generate, to prevent collision with the PDF output from other plugins that are echo or print_r information before.
* Deprecated WooCommerce hook woocommerce_add_order_item_meta is removed
* With bulk operation the PDF ticket attachments are cleaned up before processing the next order

= 1.3.15 - 2023-04-23 =
* New option on the product to make the name per ticket mandatory

= 1.3.14 - 2023-04-14 =
* Fix output of name counter on ticket
* Fix URL path for the ticket detail view
* New options (including label) to display the registered user value on the ticket

= 1.3.13 - 2023-04-12 =
* New option to define the ticket detail page path to display the ticket detail
* Download link and ticket number on the email is now more visible
* New webhook for usage of restricted code on order item

= 1.3.12 - 2023-04-10 =
* Distract free ticket scanner option
* Force responsive for the ticket scanner view for better experience
* Fix retrieving default value of an option

= 1.3.11 - 2023-04-05 =
* Fix validation meta data object for the ticket
* Add new translation option for the ticket heading
* Small performance improvement getting the codes for the backend
* Added some checks for broken codes while retrieving them for the admin area

= 1.3.10 - 2023-04-03 =
* Fix for the ticket scanner access control description text
* Add a new QR image in the backend to have the QR code with the public ticket number, that can be also used with the ticket scanner
* WARNING: Default option value changed for "Clear the ticket number if the order was deleted, canceled or refunded"! The default is now: activated - what makes more sense

= 1.3.9 - 2023-03-25 =
* Fix for pages without correct timezone setting

= 1.3.8 - 2023-03-24 =
* Support for old WC added

= 1.3.7 - 2023-03-23 =
* Fix name per ticket display
* Change the layout and heading of ordered items on the ticket details
* Expiration warning on the ticket fixed for events without a end date

= 1.3.6 - 2023-03-21 =
* Remove the name per ticket from the product meta and moved it to the session
* Session data for the name per ticket will be deleted after the purchase

= 1.3.5 - 2023-03-20 =
* Fix column customer if the order was deleted already
* New option on the product to collect names for each sold ticket. The name will be printed on the ticket
* Added new export column for the public ticket id
* Your server timezone is used on the ticket scanner to determine the expiration status
* New option to print all qr codes of the tickets for a product

= 1.3.4 - 2023-03-09 =
* Access to admin area can be controlled by roles
* New option to prevent ticket beeing too early redeemed
* Ticket scanner has now an option to automatically redeem the ticket after retrieving it

= 1.3.3 - 2023-03-02 =
* New option for variable products, you can exclude now variants to be treated like a ticket
* Download option for the ICS calendar file within the product detail view
* Adjusted the ICS File to use the timezone
* Adjusted the ICS File - if no start time is set, it will be a day event
* If the order was not paid or completet, then the payment order dates are not shown on the ticket
* Fix for the missing end time on the ticket EXPIRE message

= 1.3.2 - 2023-02-23 =
* New option to display the ticket list name on the ticket
* New option to display the ticket list description on the ticket
* New option to display all purchased items of the order on the ticket
* If the customer order note is emtpy, the quotes are nor displayed
* Option area reposition some option to make it easier to find them
* Optimize the PDF invoice print - still a problem if you have so many ticket numbers, that they cannot fit within one invoice page
* Added optimization for plugin "PDF Invoices & Packing Slips for WooCommerce" and many tickets sold in one order
* Delete ticket repaired

= 1.3.1 - 2023-02-20 =
* Fix ticket list delete and tickets
* Add new option to have the purchase customer note on the ticket

= 1.3.0 - 2023-02-15 =
* Added support for plain permalinks for the ticket scanner
* New option to hide the event date from the ticket
* New option to activate a compatibility mode for the ticket urls in case you experience issues
* Fixed color issue of the ticket title and event name, so that your theme is not overwriting it

sasoEventtickets_Core.php
<?php
include(plugin_dir_path(__FILE__)."init_file.php");
class sasoEventtickets_Core {
	private $MAIN;

	private $_CACHE_list = [];

	public $ticket_url_path_part = "ticket";

	public function __construct($MAIN) {
		if ($MAIN->getDB() == null) throw new Exception("#9999 DB needed");
		$this->MAIN = $MAIN;
	}

	private function getBase() {
		return $this->MAIN->getBase();
	}
	private function getDB() {
		return $this->MAIN->getDB();
	}

	public function clearCode($code) {
		return str_replace(" ","",str_replace(":","",str_replace("-", "", $code)));
	}

	public function getListById($id) {
		$sql = "select * from ".$this->getDB()->getTabelle("lists")." where id = ".intval($id);
		$ret = $this->getDB()->_db_datenholen($sql);
		if (count($ret) == 0) throw new Exception("#9232 not found");
		return $ret[0];
	}

	public function getCodesByRegUserId($user_id) {
		$user_id = intval($user_id);
		if ($user_id <= 0) return [];
		$sql = "select a.* from ".$this->getDB()->getTabelle("codes")." a where user_id = ".$user_id;
		return $this->getDB()->_db_datenholen($sql);
	}

	public function retrieveCodeByCode($code, $mitListe=false) {
		$code = $this->clearCode($code);
		$code = $this->getDB()->reinigen_in($code);
		if (empty($code)) throw new Exception("#203 code empty");
		if ($mitListe) {
			$sql = "select a.*, b.name as list_name from ".$this->getDB()->getTabelle("codes")." a
					left join ".$this->getDB()->getTabelle("lists")." b on a.list_id = b.id
					where code = '".$code."'";
		} else {
			$sql = "select a.* from ".$this->getDB()->getTabelle("codes")." a where code = '".$code."'";
		}
		$ret = $this->getDB()->_db_datenholen($sql);
		if (count($ret) == 0) throw new Exception("#204 code not found");
		return $ret[0];
	}

	public function checkCodesSize() {
		if ($this->isCodeSizeExceeded()) throw new Exception("#208 too many codes. Unlimited codes only with premium");
	}
	public function isCodeSizeExceeded() {
		return $this->getBase()->premiumCheck_isAllowedAddingCode($this->getDB()->getCodesSize()) == false;
	}

	public function retrieveCodeById($id, $mitListe=false) {
		$id = intval($id);
		if ($id == 0) throw new Exception("#220 id of code empty");
		if ($mitListe) {
			$sql = "select a.*, b.name as list_name from ".$this->getDB()->getTabelle("codes")." a
					left join ".$this->getDB()->getTabelle("lists")." b on a.list_id = b.id
					where a.id = ".$id;
		} else {
			$sql = "select a.* from ".$this->getDB()->getTabelle("codes")." a where a.id = ".$id;
		}
		$ret = $this->getDB()->_db_datenholen($sql);
		if (count($ret) == 0) throw new Exception("#221 code not found");
		return $ret[0];
	}

	public function getMetaObject() {
		$metaObj = [
			'validation'=>['first_success'=>'', 'first_ip'=>'', 'last_success'=>'', 'last_ip'=>'']
			,'user'=>[
				'reg_approved'=>0,
				'reg_request'=>'',
				'value'=>'',
				'reg_ip'=>'',
				'reg_userid'=>0, '
				_reg_username'=>'']
			,'used'=>[
				'reg_ip'=>'',
				'reg_request'=>'',
				'reg_userid'=>0,
				'_reg_username'=>'']
			,'confirmedCount'=>0
			,'woocommerce'=>[
					'order_id'=>0, 'product_id'=>0,
					'creation_date'=>0, 'item_id'=>0,
					'user_id'=>0] // product code for sale
			,'wc_rp'=>['order_id'=>0, 'product_id'=>0, 'creation_date'=>0, 'item_id'=>0] // restriction purchase used
			,'wc_ticket'=>[
					'is_ticket'=>0,
					'ip'=>'',
					'userid'=>0,
					'_username'=>'',
					'redeemed_date'=>'',
					'redeemed_by_admin'=>0,
					'set_by_admin'=>0,
					'set_by_admin_date'=>'',
					'idcode'=>'',
					'_url'=>'',
					'_public_ticket_id'=>'',
					'stats_redeemed'=>[],
					'name_per_ticket'=>''
					] // ticket purchase ; stats_redeemed is only used if the ticket can be redeemed more than once
			];

		if ($this->MAIN->isPremium() && method_exists($this->MAIN->getPremiumFunctions(), 'getMetaObject')) {
			$metaObj = $this->MAIN->getPremiumFunctions()->getMetaObject($metaObj);
		}

		return $metaObj;
	}
	public function encodeMetaValuesAndFillObject($metaValuesString, $codeObj=null) {
		$metaObj = $this->getMetaObject();
		if (!empty($metaValuesString)) {
			$metaObj = array_replace_recursive($metaObj, json_decode($metaValuesString, true));
		}
		if (isset($metaObj['user']['reg_userid']) && $metaObj['user']['reg_userid'] > 0) {
			$u = get_userdata($metaObj['user']['reg_userid']);
			if ($u === false) {
				$metaObj['user']['_reg_username'] = esc_html__("USERID DO NOT EXISTS", 'event-tickets-with-ticket-scanner');
			} else {
				$metaObj['user']['_reg_username'] = $u->first_name." ".$u->last_name." (".$u->user_login.")";
			}
		} else {
			$metaObj['user']['_reg_username'] = "";
		}
		if (isset($metaObj['used']['reg_userid']) && $metaObj['used']['reg_userid'] > 0) {
			$u = get_userdata($metaObj['used']['reg_userid']);
			if ($u === false) {
				$metaObj['used']['_reg_username'] = esc_html__("USERID DO NOT EXISTS", 'event-tickets-with-ticket-scanner');
			} else {
				$metaObj['used']['_reg_username'] = $u->first_name." ".$u->last_name." (".$u->user_login.")";
			}
		} else {
			$metaObj['used']['_reg_username'] = "";
		}
		if (isset($metaObj['wc_ticket']['userid']) && $metaObj['wc_ticket']['userid'] > 0) {
			$u = get_userdata($metaObj['wc_ticket']['userid']);
			if ($u === false) {
				$metaObj['wc_ticket']['_username'] = esc_html__("USERID DO NOT EXISTS", 'event-tickets-with-ticket-scanner');
			} else {
				$metaObj['wc_ticket']['_username'] = $u->first_name." ".$u->last_name." (".$u->user_login.")";
			}
		} else {
			$metaObj['wc_ticket']['_username'] = "";
		}
		if (isset($metaObj['wc_ticket']['redeemed_by_admin']) && $metaObj['wc_ticket']['redeemed_by_admin'] > 0) {
			$u = get_userdata($metaObj['wc_ticket']['redeemed_by_admin']);
			if ($u === false) {
				$metaObj['wc_ticket']['_redeemed_by_admin_username'] = esc_html__("USERID DO NOT EXISTS", 'event-tickets-with-ticket-scanner');
			} else {
				$metaObj['wc_ticket']['_redeemed_by_admin_username'] = $u->first_name." ".$u->last_name." (".$u->user_login.")";
			}
		} else {
			$metaObj['wc_ticket']['_redeemed_by_admin_username'] = "";
		}
		if (isset($metaObj['wc_ticket']['set_by_admin']) && $metaObj['wc_ticket']['set_by_admin'] > 0) {
			$u = get_userdata($metaObj['wc_ticket']['set_by_admin']);
			if ($u === false) {
				$metaObj['wc_ticket']['_set_by_admin_username'] = esc_html__("USERID DO NOT EXISTS", 'event-tickets-with-ticket-scanner');
			} else {
				$metaObj['wc_ticket']['_set_by_admin_username'] = $u->first_name." ".$u->last_name." (".$u->user_login.")";
			}
		} else {
			$metaObj['wc_ticket']['_set_by_admin_username'] = "";
		}
		if ($metaObj['wc_ticket']['is_ticket'] == 1 && $codeObj != null && is_array($codeObj)) {
			if (empty($metaObj['wc_ticket']['idcode']))	$metaObj['wc_ticket']['idcode'] = crc32($codeObj['id']."-".time());
			if (empty($metaObj['wc_ticket']['_public_ticket_id'])) $metaObj['wc_ticket']['_public_ticket_id'] = $this->getTicketId($codeObj, $metaObj);
			$metaObj['wc_ticket']['_url'] = $this->getTicketURL($codeObj, $metaObj);
		}

		// update validation fields
		if ($metaObj['confirmedCount'] > 0) {
			if (empty($metaObj['validation']['first_success'])) {
				// check used wert
				if ( !empty($metaObj['used']['reg_request']) ) {
					if (empty($metaObj['validation']['first_success'])) $metaObj['validation']['first_success'] = $metaObj['used']['reg_request'];
					if (empty($metaObj['validation']['first_ip'])) $metaObj['validation']['first_ip'] = $metaObj['used']['reg_ip'];
				} elseif (!empty($metaObj['user']['reg_request'])) { // check user reg wert
					if (empty($metaObj['validation']['first_success'])) $metaObj['validation']['first_success'] = $metaObj['user']['reg_request'];
					if (empty($metaObj['validation']['first_ip'])) $metaObj['validation']['first_ip'] = $metaObj['user']['reg_ip'];
				}
			}
		}

		if ($this->MAIN->isPremium() && method_exists($this->MAIN->getPremiumFunctions(), 'encodeMetaValuesAndFillObject')) {
			$metaObj = $this->MAIN->getPremiumFunctions()->encodeMetaValuesAndFillObject($metaObj, $codeObj);
		}
		return $metaObj;
	}

	public function getMetaObjectKeyList($metaObj, $prefix="META_") {
		$keys = [];
		$prefix = strtoupper(trim($prefix));
		foreach(array_keys($metaObj) as $key) {
			$tag = $prefix.strtoupper($key);
			if (is_array($metaObj[$key])) {
				$_keys = $this->getMetaObjectKeyList($metaObj[$key], $tag."_");
				$keys = array_merge($keys, $_keys);
			} else {
				$keys[] = $tag;
			}
		}
		return $keys;
	}

	public function getMetaObjectAllowedReplacementTags() {
		$tags = [];
		$allowed_tags = [
			"USER_VALUE"=>esc_html__("Value given by the user during the code registration.", 'event-tickets-with-ticket-scanner'),
			"USER_REG_IP"=>esc_html__("IP address of the user, register to a code.", 'event-tickets-with-ticket-scanner'),
			"USER_REG_USERID"=>esc_html__("User id of the registered user to a code. Default will be 0.", 'event-tickets-with-ticket-scanner'),
			"USED_REG_IP"=>esc_html__("IP addres of the user that used the code.", 'event-tickets-with-ticket-scanner'),
			"CONFIRMEDCOUNT"=>esc_html__("Amount of how many times the code was validated successfully.", 'event-tickets-with-ticket-scanner'),
			"WOOCOMMERCE_ORDER_ID"=>esc_html__("WooCommerce order id assigned to the code.", 'event-tickets-with-ticket-scanner'),
			"WOOCOMMERCE_PRODUCT_ID"=>esc_html__("WooCommerce product id assigned to the code.", 'event-tickets-with-ticket-scanner'),
			"WOOCOMMERCE_CREATION_DATE"=>esc_html__("Creation date of the WooCommerce sales date.", 'event-tickets-with-ticket-scanner'),
			"WOOCOMMERCE_USER_ID"=>esc_html__("User id of the WooCommerce sales.", 'event-tickets-with-ticket-scanner'),
			"WC_RP_ORDER_ID"=>esc_html__("WooCommerce order id, that was purchases using this code as an allowance to purchase a restricted product.", 'event-tickets-with-ticket-scanner'),
			"WC_RP_PRODUCT_ID"=>esc_html__("WooCommerce product id that was restricted with this code.", 'event-tickets-with-ticket-scanner'),
			"WC_RP_CREATION_DATE"=>esc_html__("Creation date of the WooCommerce purchase using the allowance code.", 'event-tickets-with-ticket-scanner')
		];
		foreach($allowed_tags as $key => $value) {
			$tags[] = ["key"=>$key, "label"=>$value];
		}
		return $tags;
	}

	// returns a default meta object for a code list
	public function getMetaObjectList() {
		$metaObj = [
			'desc'=>'',
			'redirect'=>['url'=>''],
			'formatter'=>[
				'active'=>1,
				'format'=>'' // JSON mit den Format Werten
				]
		];
		if ($this->MAIN->isPremium() && method_exists($this->MAIN->getPremiumFunctions(), 'getMetaObjectList')) {
			$metaObj = $this->MAIN->getPremiumFunctions()->getMetaObjectList($metaObj);
		}
		return $metaObj;
	}

	public function encodeMetaValuesAndFillObjectList($metaValuesString) {
		$metaObj = $this->getMetaObjectList();
		if (!empty($metaValuesString)) {
			$metaObj = array_replace_recursive($metaObj, json_decode($metaValuesString, true));
		}
		return $metaObj;
	}

	public function json_encode_with_error_handling($object) {
		$json = json_encode($object, JSON_NUMERIC_CHECK);
		if (json_last_error() !== JSON_ERROR_NONE) {
			throw new Exception(json_last_error_msg());
		}
		return $json;
	}

	public function getRealIpAddr() {
	    if (!empty($_SERVER['HTTP_CLIENT_IP']))   //check ip from share internet
	    {
	      $ip=sanitize_text_field($_SERVER['HTTP_CLIENT_IP']);
	    }
	    elseif (!empty($_SERVER['HTTP_X_FORWARDED_FOR']))   //to check ip is pass from proxy
	    {
	      $ip=sanitize_text_field($_SERVER['HTTP_X_FORWARDED_FOR']);
	    }
	    else
	    {
	      $ip=sanitize_text_field($_SERVER['REMOTE_ADDR']);
	    }
	    return $ip;
	}

	public function triggerWebhooks($status, $codeObj) {
		$options = $this->MAIN->getOptions();
		if ($options->isOptionCheckboxActive('webhooksActiv')) {
			$optionname = "";
			switch($status) {
				case 0:
					$optionname = "webhookURLinvalid";
					break;
				case 1:
					$optionname = "webhookURLvalid";
					break;
				case 2:
					$optionname = "webhookURLinactive";
					break;
				case 3:
					$optionname = "webhookURLisregistered";
					break;
				case 4:
					$optionname = "webhookURLexpired";
					break;
				case 5:
					$optionname = "webhookURLmarkedused";
					break;
				case 6:
					$optionname = "webhookURLsetused";
					break;
				case 7:
					$optionname = "webhookURLregister";
					break;
				case 8:
					$optionname = "webhookURLipblocking";
					break;
				case 9:
					$optionname = "webhookURLipblocked";
					break;
				case 10:
					$optionname = "webhookURLaddwcinfotocode";
					break;
				case 11:
					$optionname = "webhookURLwcremove";
					break;
				case 12:
					$optionname = "webhookURLaddwcticketinfoset";
					break;
				case 13:
					$optionname = "webhookURLaddwcticketredeemed";
					break;
				case 14:
					$optionname = "webhookURLaddwcticketunredeemed";
					break;
				case 15:
					$optionname = "webhookURLaddwcticketinforemoved";
					break;
				case 16:
					$optionname = "webhookURLrestrictioncodeused";
					break;
			}
			if (!empty($optionname)) {
				$url = $options->getOption($optionname)['value'];
				if (!empty($url)) {
					$url = $this->replaceURLParameters($url, $codeObj);
					wp_remote_get($url);
				}
			}
		}
	}

	private function _getCachedList($list_id) {
		if (isset($this->_CACHE_list[$list_id])) return $this->_CACHE_list[$list_id];
		$this->_CACHE_list[$list_id] = $this->getListById($list_id);
		return $this->_CACHE_list[$list_id];
	}

	public function replaceURLParameters($url, $codeObj) {
		$url = str_replace("{CODE}", isset($codeObj['code']) ? $codeObj['code'] : '', $url);
		$url = str_replace("{CODEDISPLAY}", isset($codeObj['code_display']) ? $codeObj['code_display'] : '', $url);
		$url = str_replace("{IP}", $this->getRealIpAddr(), $url);
		$userid = '';
		if (is_user_logged_in()) {
			$userid = get_current_user_id();
		}
		$url = str_replace("{USERID}", $userid, $url);

		$listname = "";
		if (isset($codeObj['list_id']) && $codeObj['list_id'] > 0 && strpos($url, "{LIST}") !== false) {
			try {
				$listObj = $this->_getCachedList($codeObj['list_id']);
				$listname = $listObj['name'];
			} catch (Exception $e) {
			}
		}
		$url = str_replace("{LIST}", urlencode($listname), $url);

		$listdesc = "";
		if (isset($codeObj['list_id']) && $codeObj['list_id'] > 0 && strpos($url, "{LIST_DESC}") !== false) {
			try {
				$listObj = $this->_getCachedList($codeObj['list_id']);
				$metaObj = [];
				if (!empty($listObj['meta'])) $metaObj = $this->encodeMetaValuesAndFillObjectList($listObj['meta']);
				if (isset($metaObj['desc'])) $listdesc = $metaObj['desc'];
			} catch (Exception $e) {
			}
		}
		$url = str_replace("{LIST_DESC}", urlencode($listdesc), $url);

		$metaObj = [];
		if (!empty($codeObj['meta'])) $metaObj = $this->encodeMetaValuesAndFillObject($codeObj['meta'], $codeObj);
		if (count($metaObj) > 0) $url = $this->_replaceTagsInTextWithMetaObjectsValues($url, $metaObj, "META_");

		return $url;
	}

	private function _replaceTagsInTextWithMetaObjectsValues($text, $metaObj, $prefix="") {
		$prefix = strtoupper(trim($prefix));
		foreach(array_keys($metaObj) as $key) {
			$tag = $prefix.strtoupper($key);
			if (is_array($metaObj[$key])) {
				$text = $this->_replaceTagsInTextWithMetaObjectsValues($text, $metaObj[$key], $tag."_");
			} else {
				$text = str_replace("{".$tag."}", urlencode($metaObj[$key]), $text);
			}
		}
		return $text;
	}

	public function checkCodeExpired($codeObj) {
		if ($this->MAIN->isPremium() && method_exists($this->MAIN->getPremiumFunctions(), 'checkCodeExpired')) {
			if ($this->MAIN->getPremiumFunctions()->checkCodeExpired($codeObj)) {
				return true;
			}
		}
		return false;
	}
	public function isCodeIsRegistered($codeObj) {
		$meta = [];
		if (!empty($codeObj['meta'])) $meta = $this->encodeMetaValuesAndFillObject($codeObj['meta'], $codeObj);
		if (isset($meta['user']) && isset($meta['user']['value']) && !empty($meta['user']['value'])) {
			return true;
		}
		return false;
	}

	public function getTicketURLBase($defaultPath=false) {
		$path = plugin_dir_url(__FILE__).$this->ticket_url_path_part;
		if ($defaultPath == false) {
			$wcTicketCompatibilityModeURLPath = trim($this->MAIN->getOptions()->getOptionValue('wcTicketCompatibilityModeURLPath'));
			$wcTicketCompatibilityModeURLPath = trim(trim($wcTicketCompatibilityModeURLPath, "/"));
			if (!empty($wcTicketCompatibilityModeURLPath)) {
				$path = site_url()."/".$wcTicketCompatibilityModeURLPath;
			}
		}
		return $path."/";
	}
	public function getTicketId($codeObj, $metaObj) {
		if (isset($codeObj['code']) && isset($codeObj['order_id']) && isset($metaObj['wc_ticket']['idcode'])) {
			return $metaObj['wc_ticket']['idcode']."-".$codeObj['order_id']."-".$codeObj['code'];
		} else {
			return "";
		}
	}
	public function getTicketURL($codeObj, $metaObj) {
		$ticket_id = $this->getTicketId($codeObj, $metaObj);
		$baseURL = $this->getTicketURLBase();
		$url = $baseURL.$ticket_id;
		if ($this->MAIN->getOptions()->isOptionCheckboxActive('wcTicketCompatibilityMode')) {
			$url = $baseURL."?code=".$ticket_id;
		}
		return $url;
	}
	public function getOrderTicketsURL($order) {
		if ($order == null) throw new Exception("Order empty - no order tickets PDF url created");
		$order_id = $order->get_id();
		$idcode = $order->get_meta('_saso_eventtickets_order_idcode');
		if (empty($idcode)) {
			$idcode = strtoupper(md5($order_id."-".time()."-".uniqid()));
			$order->update_meta_data( '_saso_eventtickets_order_idcode', $idcode );
			$order->save();
		}
		$baseURL = $this->getTicketURLBase();
		$ticket_id = "order-".$order_id."-".$idcode;
		$url = $baseURL.$ticket_id;
		if ($this->MAIN->getOptions()->isOptionCheckboxActive('wcTicketCompatibilityMode')) {
			$url = $baseURL."?code=".$ticket_id;
		}
		return $url;
	}
	public function getTicketURLPath($defaultPath=false) {
		$p = $this->getTicketURLBase($defaultPath);
		$teile = parse_url($p);
		return $teile['path'];
	}
	public function getTicketURLComponents($url) {
		$teile = explode("/", $url);
		$teile = array_reverse($teile);
		$ret = "";
		$request = "";
		$is_pdf_request = false;
		$is_ics_request = false;
		$is_badge_request = false;
		foreach($teile as $teil) {
			$teil = trim($teil);
			if (empty($teil)) continue;
			if (strtolower($teil) == "?pdf") continue;
			if (strtolower($teil) == "?ics") continue;
			if ($teil == $this->ticket_url_path_part) break;
			$ret = $teil;
			break;
		}
		if (isset($_GET['code'])) {
			$parts = explode("-", trim($_GET['code']));
			$t = explode("?", $url);
			if (count($t) > 1) {
				unset($t[0]);
				$request = join("&", $t);
			}
			$is_pdf_request = in_array("pdf", $t);
			$is_ics_request = in_array("ics", $t);
			$is_badge_request = in_array("badge", $t);
		} else {
			if (empty($ret)) throw new Exception("#9301 ticket id not found");
			$parts = explode("-", $ret);
			$t = explode("?", $parts[2]);
			$parts[2] = $t[0];
			if (count($t) > 1) {
				unset($t[0]);
				$request = join("&", $t);
			}
			$is_pdf_request = in_array("pdf", $t) || isset($_GET['pdf']);
			$is_ics_request = in_array("ics", $t) || isset($_GET['ics']);
			$is_badge_request = in_array("badge", $t) || isset($_GET['badge']);
		}
		if (count($parts) != 3) throw new Exception("#9302 ticket id not correct");
		$parts[2] = str_replace("?pdf", "", $parts[2]);
		$parts[2] = str_replace("?ics", "", $parts[2]);
		$parts_assoc = [
			"idcode"=>$parts[0],
			"order_id"=>$parts[1],
			"code"=>$parts[2],
			"_request"=>$request,
			"_isPDFRequest"=>$is_pdf_request,
			"_isICSRequest"=>$is_ics_request,
			"_isBadgeRequest"=>$is_badge_request
		];
		return $parts_assoc;
	}

	public function mergePDFs($filepaths, $filename, $filemode="I", $deleteFilesAfterMerge=true) {
		if (count($filepaths) > 0) {
			if (!class_exists('sasoEventtickets_PDF')) {
				require_once("sasoEventtickets_PDF.php");
			}
			$pdf = new sasoEventtickets_PDF();
			$pdf->setFilemode($filemode);
			$pdf->setFilename($filename);
			try {
				$pdf->mergeFiles($filepaths); // send file to browser if,filemode is I
			} catch(Exception $e) {}

			// clean up temp files
			if ($deleteFilesAfterMerge) {
				foreach($filepaths as $filepath) {
					if (file_exists($filepath)) {
						@unlink($filepath);
					}
				}
			}
			if ($pdf->getFilemode() == "F") {
				return $pdf->getFullFilePath();
			} else {
				exit;
			}
		}
	}

	public function my_upgrade_function( $upgrader_object, $options ) {
    	$current_plugin_path_name = plugin_basename( __FILE__ );
    	if ($options['action'] == 'update' && $options['type'] == 'plugin' ) {
			if (isset($options['plugins'])) {
				foreach($options['plugins'] as $each_plugin) {
					if ($each_plugin==$current_plugin_path_name) {
					// .......................... YOUR CODES .............

					}
				}
			}
    	}
	}

}
?>